### PR TITLE
Change local charm deploy to updated series selection logic.

### PR DESCRIPTION
**Pull request by ExternalReality**
## Description of change

There are two reasons why this change is needed:

1. To fix the bug referenced below
2. To make the behavior of deploying charms more consistent (see below)

#### Bootstrap a controller
```sh
juju bootstrap lxd bug_1663510
```

#### Ensure that the model's default series is set to `xenial`

```sh
juju model-config default-series=xenial
```

#### Download promethus charm archive
```sh
wget https://api.jujucharms.com/charmstore/v5/~prometheus-charmers/prometheus-14/archive
```
#### Unzip the archive to a local directory (not necessary but we aim to match the bug's scenario as closely as possible) 
```sh
unzip archive -d prometheus
```

#### Deploy the local prometheus charm
```sh
juju deploy ./prometheus
```

The deployed series should be `xenial` and **not** `trusty`

## Documentation changes

This change **does** affect the CLI in that whenever you use the `--series=<series>` option to specify a series when deploying a **local charm**, you must now use the `--force` flag if the series is not listed as a supported charm in the charm's `metadata.yml` file. This is the behavior of other commands which use the `--series` option.

## Bug reference

[lp#1663510](https://bugs.launchpad.net/juju/+bug/1663510)

### CODE REVIEW

**Comment by ExternalReality** [comment]
Tests were updated to use `--force` flag when calling deploy because the test charm files do not specify supported series. (see Documentation Changes)

File: cmd/juju/application/deploy.go [hunk 1]
```diff
@@ -794,9 +794,54 @@ func (c *DeployCommand) maybeReadLocalBundle() (deployFn, error) {
...
+	return false
+}
+
+func (c *DeployCommand) maybeReadLocalCharm(apiRoot DeployAPI) (deployFn, error) {
+	//NOTE: Here we select the series using the algorithm defined by the
+	//series selector. This serves to override the algorithm found in
+	//`charmrepo.NewCharmAtPath` which is outdated (but must still be called
+	//since the code is coupled with path interpretation logic which needs
+	//to be factored out).
+
```
Comment (ExternalReality) [inline cmd/juju/application/deploy.go]: IMHO, the `seriesSelector` should be moved into the charmrepo package as part of a refactoring of the package. For example the `charmrepo.NewCharmAtPathForceSeries` both serves to determine if the charm reference can be interpreted as a local charm path and also determines its series. The existence of the `seriesSelector` illustrates the need to have these concerns separate.

File: cmd/juju/application/deploy.go [hunk 2]
```diff
@@ -663,7 +663,7 @@ func (c *DeployCommand) Run(ctx *cmd.Context) error {
 
 	deploy, err := findDeployerFIFO(
 		c.maybeReadLocalBundle,
-		c.maybeReadLocalCharm,
+		func() (deployFn, error) { return c.maybeReadLocalCharm(apiRoot) },
```
Comment (ExternalReality) [inline cmd/juju/application/deploy.go]: I used a lambda here but it could be done like 668 below.

File: cmd/juju/application/deploy.go [hunk 3]
```diff
@@ -794,9 +794,41 @@ func (c *DeployCommand) maybeReadLocalBundle() (deployFn, error) {
 	}, nil
 }
 
-func (c *DeployCommand) maybeReadLocalCharm() (deployFn, error) {
+func (c *DeployCommand) maybeReadLocalCharm(apiRoot DeployAPI) (deployFn, error) {
+	//NOTE: Here we select the series using the algorithm defined by the
```
Comment (howbazaar) [inline cmd/juju/application/deploy.go]: NIT: single space after the // for comments.

File: cmd/juju/application/deploy.go [hunk 4]
```diff
@@ -794,9 +794,41 @@ func (c *DeployCommand) maybeReadLocalBundle() (deployFn, error) {
...
+	//series selector. This serves to override the algorithm found in
+	//`charmrepo.NewCharmAtPath` which is outdated (but must still be called
+	//since the code is coupled with path interpretation logic which needs
+	//to be factored out).
+
+	ch, err := charm.ReadCharm(c.CharmOrBundle)
+	var series string
+	if err != nil {
```
Comment (howbazaar) [inline cmd/juju/application/deploy.go]: It seems that we are saying "if you can't read the charm or bundle, use the series specified by the CLI"

But if we can't read the charm or bundle, is it worthwhile continuing at all?

File: cmd/juju/application/deploy.go [hunk 5]
```diff
@@ -794,9 +794,41 @@ func (c *DeployCommand) maybeReadLocalBundle() (deployFn, error) {
...
+			force:           c.Force,
+			conf:            modelCfg,
+			fromBundle:      false,
+		}
+
+		series, err = seriesSelector.charmSeries()
+		if charm.IsUnsupportedSeriesError(err) {
+			return nil, errors.Errorf("%v. Use --force to deploy the charm anyway.", err)
+		} else if err != nil {
+			series = c.Series
```
Comment (howbazaar) [inline cmd/juju/application/deploy.go]: What is the use case here? When will charmSeries return an error that isn't an unsupported series error?

Does it still make sense to just use the CLI --series, or should we be telling the user?

File: cmd/juju/application/deploy_test.go [hunk 6]
```diff
@@ -138,7 +138,7 @@ func (s *DeploySuite) TestPathWithNoCharmOrBundle(c *gc.C) {
 
 func (s *DeploySuite) TestCharmDir(c *gc.C) {
 	ch := testcharms.Repo.ClonedDirPath(s.CharmsPath, "dummy")
-	err := runDeploy(c, ch, "--series", "trusty")
+	err := runDeploy(c, ch, "--series", "trusty", "--force")
```
Comment (howbazaar) [inline cmd/juju/application/deploy_test.go]: Wouldn't it just be easier to change our testing charms to be multi-series for the series that we use them for?

File: cmd/juju/application/deploy.go [hunk 7]
```diff
@@ -794,9 +794,41 @@ func (c *DeployCommand) maybeReadLocalBundle() (deployFn, error) {
...
+	//NOTE: Here we select the series using the algorithm defined by the
+	//series selector. This serves to override the algorithm found in
+	//`charmrepo.NewCharmAtPath` which is outdated (but must still be called
+	//since the code is coupled with path interpretation logic which needs
+	//to be factored out).
+
+	ch, err := charm.ReadCharm(c.CharmOrBundle)
+	var series string
+	if err != nil {
+		series = c.Series
```
Comment (ExternalReality) [inline cmd/juju/application/deploy.go]: Reading the charm here is only meant to aid in inferring the correct series, if this fails we just fall back to the argument series. If reading the charm fails here is will also fail below were it is handled properly. This is just an expedient to get the correct series. A proper refactoring of the `charmrepo` package is needed for a more elegant fix.

**Comment by nskaggs** [comment]
!!build!!

**Comment by ExternalReality** [comment]
!!build!!

**Comment by ExternalReality** [comment]
The "--force" flag has been removed for every place where I think it makes sense to remove it (90% of the tests where the flag was added). It was added to those test to make them compliant with the rules of the `seriesSelector` abstraction but we can were able to work around needing it by using multi-series test charms where applicable.

**Comment by ExternalReality** [comment]
!!build!!

File: cmd/juju/application/deploy.go (hunk 7 above)
Comment (howbazaar) [inline cmd/juju/application/deploy.go]: This comment would be very useful in the actual code :)

File: testcharms/charm-repo/quantal/metered-multi-series/metadata.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+name: metered-multi-series
+summary: "A metered charm with custom metrics"
+description: ""
+series:
+  - xenial
```
Comment (howbazaar) [inline testcharms/charm-repo/quantal/metered-multi-series/metadata.yaml]: If we are saying it is multi-series, perhpas pub multiple series here? Add trusty.

**Comment by ExternalReality** [comment]
$$MERGE$$

**Pull request merged**